### ISSUE TITLE
Be `ionice` on unix

### ISSUE THREAD

**Issue by richard-uk1** [rust-lang/cargo#5569]
When building on unix, I don't think there would be any harm setting the i/o nice value, and it may improve UX in some circumstances. From [rustc-perf#232](https://github.com/rust-lang-nursery/rustc-perf/issues/232#issuecomment-391458003).

**Edit** The original post meant specifically the i/o scheduling nice, not the general process nice.

**Comment by derekdreery** [comment]
I'm happy to implement this, just want to see if there were any reasons it isn't a good idea first.

**Comment by alexcrichton** [comment]
I don't think that we'll want this to be the default because it runs the risk of making builds slower by default, but having it as a configuration option seems reasonable

**Comment by lukaslueg** [comment]
This should at least be behind a config-option so the default behavior remains the expected behavior, which is niceness 0. At least on Linux, you are always able to automatically move cargo into its own control group and set its CPU-, RAM-, and IO-priority. Sure it's more complicated and not built-in, yet there is a strong case to be made that `cargo` is first and foremost a process like any other; and the cgroup-solution is strictly more powerful & beyond reach by cargo.

**Comment by matklad** [comment]
>At least on Linux, you are always able to automatically move cargo into its own control group and set its CPU-, RAM-, and IO-priorit

Interesting! Perhaps we should add instructions of how to do that to Cargo docs? I feel that "I run cargo build on my dev laptop while also using a browser and IDE" is a common use-case, so many people might want to reduce Cargo's priority, even if they don't realize that it is possible.

**Comment by lukaslueg** [comment]
@matklad, let me actually check on it. There once was a control-groups management daemon which by now seems to have been replaced by `systemd`, which itself lacks the feature.

**Comment by hammerandtongs** [comment]
@derekdreery 

I'm the linked post from that thread.

Just to be clear I said IONICE

"""ionice - set or get process I/O scheduling class and priority"""

This is much more effective for me with less negative impact on the build.

Linux desktop io scheduling needs a bit more help then process scheduling but ymmv.

**Comment by derekdreery** [comment]
@hammerandtongs Updated.

**Comment by lukaslueg** [comment]
I've looked into this again and it seems that the daemon to place processes into control groups automatically have been deprecated without being replaced. The way to go seems to be using `systemd-run` to create a transient, resource-constrained `scope` and spawn `cargo` within. For examples

```
systemd-run --scope --slice=user.slice -p IOAccounting=true -p IOAccounting=true -p CPUWeight=1 -p IOWeight=1 cargo test
```

This creates a new `scope` under `user.slice` with minimal IO- and CPU-weights. That is, the `user-1000`-slice (or whatever your "real" users runs under) and all it's processes will completely dominate the `cargo`-process for resources.

Relevant documentation for [`systemd-run`](https://www.freedesktop.org/software/systemd/man/systemd-run.html#) and [`systemd.resource-control`](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html#).
